File: LittleBigMouse.Daemon/LittleBigMouse.Engine/Zone.cpp
#include "pch.h"
#include "zone.h"

#include <algorithm>
#include <iterator>

#include "tinyxml2.h"
#include "XmlHelper.h"
#include "ZoneLink.h"
#include "ZonesLayout.h"

bool Zone::IsMain() const
{
    return Main == this;
}

void Zone::ComputeDpi()
{
	const double dpiX = _pixelsBounds.Width() / (_physicalBounds.Width() / 25.4);
	const double dpiY = _pixelsBounds.Height() / (_physicalBounds.Height() / 25.4);

	Dpi = sqrt(dpiX * dpiX + dpiY * dpiY) / sqrt(2);
}

geo::Point<double> Zone::ToPhysical(const geo::Point<long> px) const
{
    auto x = _physicalBounds.Left() + ((0.5 + static_cast<double>(px.X() - _pixelsBounds.Left())) * _physicalBounds.Width() / static_cast<double>(_pixelsBounds.Width()));
    auto y = _physicalBounds.Top() + ((0.5 + static_cast<double>(px.Y() - _pixelsBounds.Top())) * _physicalBounds.Height() / static_cast<double>(_pixelsBounds.Height()));

    return {x,y};
}

geo::Point<long> Zone::ToPixels(const geo::Point<double> mm) const
{
	const auto x = _pixelsBounds.Left() + static_cast<long>(((mm.X() - _physicalBounds.Left()) * static_cast<double>(_pixelsBounds.Width()) / _physicalBounds.Width()));
	const auto y = _pixelsBounds.Top() + static_cast<long>(((mm.Y() - _physicalBounds.Top()) * static_cast<double>(_pixelsBounds.Height()) / _physicalBounds.Height()));

    return {x,y};
}

geo::Point<long> Zone::CenterPixel() const
{
    auto x = _pixelsBounds.Left() + _pixelsBounds.Right() / 2;
    auto y = _pixelsBounds.Top() + _pixelsBounds.Bottom() / 2;

    return {x,y};
}

bool Zone::Contains(const geo::Point<long>& pixel) const
{
    if (pixel.X() < _pixelsBounds.Left()) return false;
    if (pixel.Y() < _pixelsBounds.Top()) return false;
    if (pixel.X() >= _pixelsBounds.Right()) return false;
    if (pixel.Y() >= _pixelsBounds.Bottom()) return false;
    return true;
}

bool Zone::Contains(const geo::Point<double>& mm) const
{
    return _physicalBounds.Contains(mm);
}

geo::Point<long> Zone::InsidePixelsBounds(const geo::Point<long> px) const
{
    auto x = px.X();
    auto y = px.X();

    if (x < _pixelsBounds.Left()) x = _pixelsBounds.Left();
    else if (x > _pixelsBounds.Right() - 1) x = _pixelsBounds.Right() - 1;

    if (y < _pixelsBounds.Top()) y = _pixelsBounds.Top();
    else if (y > _pixelsBounds.Bottom() - 1) y = _pixelsBounds.Bottom() - 1;

    return {x,y};

}

geo::Point<double> Zone::InsidePhysicalBounds(const geo::Point<double> mm) const
{
    double x = mm.X();
    double y = mm.Y();

    if (x < _physicalBounds.Left()) x = _physicalBounds.Left();
    else if (x > _physicalBounds.Right()) x = _physicalBounds.Right();

    if (y < _physicalBounds.Top()) y = _physicalBounds.Top();
    else if (y > _physicalBounds.Bottom()) y = _physicalBounds.Bottom();

    return {x,y};
}

std::vector<geo::Rect<long>>& Zone::TravelPixels(const std::vector<Zone*>& zones, const Zone* target)
{
	const auto it = _travels.find(target->Main);
    if(it != _travels.end())
	{
        return it->second;
    }

	const std::vector<geo::Rect<long>> l = GetTravelPixels(zones, target);

    return _travels[target->Main] = l;
}

Zone::Zone(int id, std::string deviceId, std::string name, const geo::Rect<long>& pixelsBounds,
	const geo::Rect<double>& physicalBounds, Zone* main):Id(id)
	                                                     ,_pixelsBounds(pixelsBounds)
	                                                     ,_physicalBounds(physicalBounds)
	                                                     ,_physicalInside(physicalBounds)
	                                                     ,DeviceId(std::move(deviceId))
	                                                     ,Name(std::move(name))
	                                                     ,Main(main)
{
	if(!Main) Main = this;

	const double dpiX = _pixelsBounds.Width() / (_physicalBounds.Width() / 25.4);
	const double dpiY = _pixelsBounds.Height() / (_physicalBounds.Height() / 25.4);

	Dpi = sqrt(dpiX * dpiX + dpiY * dpiY) / sqrt(2);

	double pixelWidth = _physicalBounds.Width() / (double)_pixelsBounds.Width();
	double pixelHeight = _physicalBounds.Height() / (double)_pixelsBounds.Height();

	_physicalInside = geo::Rect<double> (
		_physicalBounds.Left()+pixelWidth/2,
		_physicalBounds.Top()+pixelHeight/2,
		_physicalBounds.Width()-pixelWidth,
		_physicalBounds.Height()-pixelHeight);

}


std::vector<geo::Rect<long>> Reachable(const geo::Rect<long>& source, const geo::Rect<long>& target) 
{
	const auto left = max(target.Left(),source.Left());
	const auto right = min(target.Right(),source.Right());
	const auto top = max(target.Top(),source.Top());
	const auto bottom = min(target.Bottom(),source.Bottom());

    if(left >= right) 
    {
        if(top >= bottom)
        {
            return {source};
        }

        auto start = geo::Rect<long>(source.Left(), top, source.Width(), bottom - top);
        auto dest  = geo::Rect<long>(target.Left(), top, target.Width(), bottom - top);
        return {start,dest};
    }

	if(top >= bottom) 
	{
        auto start = geo::Rect<long>(left, source.Top(), right, source.Height());
        auto dest  = geo::Rect<long>(left, target.Top(), right - left, target.Height());
		return {start,dest};
	}

    auto start = geo::Rect<long>(left, top, right-left, bottom-top);
	return {start,start};
}

std::vector<geo::Rect<long>> Travel(const geo::Rect<long>& source, const geo::Rect<long>& target, const std::vector<geo::Rect<long>>& allowed)  
{
	auto reachable = Reachable(source,target);
	if( reachable.size()>1) return reachable;

	if (allowed.empty()) return {};

	for(const auto& next: allowed)
	{
		
		std::vector<geo::Rect<long>> newAllowed;
		std::copy_if (allowed.begin(), allowed.end(), std::back_inserter(newAllowed), 
            [next](const geo::Rect<long>& value) { return 
		                      value.Left() != next.Left()
			                      && value.Top() != next.Top() 
			                      && value.Right() != next.Right() 
			                      && value.Bottom() != next.Bottom() ; 
		 });

		auto tail = Travel(next,target,newAllowed);
		if(tail.empty()) continue;

		auto travel = Travel(source,next,newAllowed);
		if(!travel.empty()) {
			travel.insert(std::end(travel), std::begin(tail), std::end(tail));
			return travel;
		}
	}

	return {};
}

std::vector<geo::Rect<long>> Zone::GetTravelPixels(const std::vector<Zone*>& zones, const Zone* target) const
{
	std::vector<geo::Rect<long>> bounds;
	for(const auto& zone: zones)
	{
        if(zone->IsMain()) bounds.push_back(zone->PixelsBounds());
    }
	return Travel(_pixelsBounds, target->PixelsBounds(), bounds);
}


void _InitZoneLinks(const ZonesLayout* layout, ZoneLink* link)
{
    while(link)
    {
		const auto id = link->TargetId;
	    for (const auto zone : layout->Zones)
		{
	        if(id == zone->Id)
	        {
		        link->Target = zone;
	            break;
	        }
	    }
	    link = link->Next;
    }
}


void Zone::InitZoneLinks(const ZonesLayout* layout) const
{
	_InitZoneLinks(layout,LeftZones);
	_InitZoneLinks(layout,TopZones);
	_InitZoneLinks(layout,RightZones);
	_InitZoneLinks(layout,BottomZones);
}


Zone::~Zone()
{
	delete LeftZones;
	delete TopZones;
	delete RightZones;
	delete BottomZones;
}

bool Zone::HorizontalReachable(const geo::Point<double>& mm) const
{
    const auto bounds = PhysicalBounds();
    const auto y = mm.Y();
	return bounds.Top() <= y && bounds.Bottom() >= y;
}

bool Zone::VerticalReachable(const geo::Point<double>& mm) const
{
    const auto bounds = PhysicalBounds();
    const auto x = mm.X();
	return bounds.Left() <= x && bounds.Right() >= x;
}

File: LittleBigMouse-Hook-Rust/src/zones/zone.rs
//! Port of `Engine/Zone`.

use super::zone_link::ZoneLink;
use super::ZoneId;
use crate::geometry::{Point, Rect};

pub struct Zone {
    pub id: i32,
    pub device_id: String,
    pub name: String,
    pixels_bounds: Rect<i32>,
    physical_bounds: Rect<f64>,
    physical_inside: Rect<f64>,
    pub dpi: f64,
    /// Self, or the `ZoneId` of an earlier zone with identical pixel bounds
    /// (a clone). C++ `Zone::Main`.
    pub main: ZoneId,
    pub left: Vec<ZoneLink>,
    pub top: Vec<ZoneLink>,
    pub right: Vec<ZoneLink>,
    pub bottom: Vec<ZoneLink>,
}

/// C++ `Zone::ComputeDpi` / the DPI computed in the constructor.
pub fn compute_dpi(pixels: Rect<i32>, physical: Rect<f64>) -> f64 {
    let dpi_x = pixels.width() as f64 / (physical.width() / 25.4);
    let dpi_y = pixels.height() as f64 / (physical.height() / 25.4);
    (dpi_x * dpi_x + dpi_y * dpi_y).sqrt() / 2.0_f64.sqrt()
}

impl Zone {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        id: i32,
        device_id: String,
        name: String,
        pixels_bounds: Rect<i32>,
        physical_bounds: Rect<f64>,
        main: ZoneId,
        left: Vec<ZoneLink>,
        top: Vec<ZoneLink>,
        right: Vec<ZoneLink>,
        bottom: Vec<ZoneLink>,
    ) -> Self {
        let dpi = compute_dpi(pixels_bounds, physical_bounds);

        let pixel_width = physical_bounds.width() / pixels_bounds.width() as f64;
        let pixel_height = physical_bounds.height() / pixels_bounds.height() as f64;
        let physical_inside = Rect::new(
            physical_bounds.left() + pixel_width / 2.0,
            physical_bounds.top() + pixel_height / 2.0,
            physical_bounds.width() - pixel_width,
            physical_bounds.height() - pixel_height,
        );

        Zone {
            id,
            device_id,
            name,
            pixels_bounds,
            physical_bounds,
            physical_inside,
            dpi,
            main,
            left,
            top,
            right,
            bottom,
        }
    }

    pub fn pixels_bounds(&self) -> Rect<i32> {
        self.pixels_bounds
    }
    pub fn physical_bounds(&self) -> Rect<f64> {
        self.physical_bounds
    }
    pub fn physical_inside(&self) -> Rect<f64> {
        self.physical_inside
    }

    /// C++ `Zone::ToPhysical`.
    pub fn to_physical(&self, px: Point<i32>) -> Point<f64> {
        let x = self.physical_bounds.left()
            + (0.5 + (px.x() - self.pixels_bounds.left()) as f64) * self.physical_bounds.width()
                / self.pixels_bounds.width() as f64;
        let y = self.physical_bounds.top()
            + (0.5 + (px.y() - self.pixels_bounds.top()) as f64) * self.physical_bounds.height()
                / self.pixels_bounds.height() as f64;
        Point::new(x, y)
    }

    /// C++ `Zone::ToPixels`.
    pub fn to_pixels(&self, mm: Point<f64>) -> Point<i32> {
        let x = self.pixels_bounds.left()
            + ((mm.x() - self.physical_bounds.left()) * self.pixels_bounds.width() as f64
                / self.physical_bounds.width()) as i32;
        let y = self.pixels_bounds.top()
            + ((mm.y() - self.physical_bounds.top()) * self.pixels_bounds.height() as f64
                / self.physical_bounds.height()) as i32;
        Point::new(x, y)
    }

    /// C++ `Zone::CenterPixel` — reproduced verbatim, including the `left +
    /// right/2` operator-precedence quirk (not a true centre).
    pub fn center_pixel(&self) -> Point<i32> {
        let x = self.pixels_bounds.left() + self.pixels_bounds.right() / 2;
        let y = self.pixels_bounds.top() + self.pixels_bounds.bottom() / 2;
        Point::new(x, y)
    }

    /// C++ `Zone::Contains(Point<long>)`.
    pub fn contains_pixel(&self, pixel: Point<i32>) -> bool {
        pixel.x() >= self.pixels_bounds.left()
            && pixel.y() >= self.pixels_bounds.top()
            && pixel.x() < self.pixels_bounds.right()
            && pixel.y() < self.pixels_bounds.bottom()
    }

    /// C++ `Zone::Contains(Point<double>)`.
    pub fn contains_mm(&self, mm: Point<f64>) -> bool {
        self.physical_bounds.contains(mm)
    }

    /// C++ `Zone::InsidePixelsBounds`.
    pub fn inside_pixels_bounds(&self, px: Point<i32>) -> Point<i32> {
        let mut x = px.x();
        let mut y = px.y();
        if x < self.pixels_bounds.left() {
            x = self.pixels_bounds.left();
        } else if x > self.pixels_bounds.right() - 1 {
            x = self.pixels_bounds.right() - 1;
        }
        if y < self.pixels_bounds.top() {
            y = self.pixels_bounds.top();
        } else if y > self.pixels_bounds.bottom() - 1 {
            y = self.pixels_bounds.bottom() - 1;
        }
        Point::new(x, y)
    }

    /// C++ `Zone::InsidePhysicalBounds`.
    pub fn inside_physical_bounds(&self, mm: Point<f64>) -> Point<f64> {
        let mut x = mm.x();
        let mut y = mm.y();
        if x < self.physical_bounds.left() {
            x = self.physical_bounds.left();
        } else if x > self.physical_bounds.right() {
            x = self.physical_bounds.right();
        }
        if y < self.physical_bounds.top() {
            y = self.physical_bounds.top();
        } else if y > self.physical_bounds.bottom() {
            y = self.physical_bounds.bottom();
        }
        Point::new(x, y)
    }

    /// C++ `Zone::HorizontalReachable`.
    pub fn horizontal_reachable(&self, mm: Point<f64>) -> bool {
        let y = mm.y();
        self.physical_bounds.top() <= y && self.physical_bounds.bottom() >= y
    }

    /// C++ `Zone::VerticalReachable`.
    pub fn vertical_reachable(&self, mm: Point<f64>) -> bool {
        let x = mm.x();
        self.physical_bounds.left() <= x && self.physical_bounds.right() >= x
    }
}
